File: python/src/strands_couchbase_memory/memory_store.py
"""Strands MemoryStore implementation backed by Couchbase Vector Search."""

from __future__ import annotations

import asyncio
import inspect
import os
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Protocol, TypedDict, cast
from uuid import uuid4

import couchbase.search as couchbase_search
from couchbase.auth import PasswordAuthenticator
from couchbase.cluster import Cluster
from couchbase.collection import Collection
from couchbase.options import ClusterOptions, SearchOptions
from couchbase.vector_search import VectorQuery, VectorSearch
from strands.memory import AddMessagesContext, MemoryEntry, MemoryStore, MemoryStoreConfig
from strands.memory import SearchOptions as StrandsSearchOptions
from strands.types.content import Message
from typing_extensions import Unpack

DEFAULT_NAME = "couchbase"
DEFAULT_DESCRIPTION = "Long-term semantic memory stored in Couchbase Vector Search."
DEFAULT_CONNECTION_STRING = "couchbase://localhost"
DEFAULT_BUCKET = "strands_memory"
DEFAULT_SCOPE = "_default"
DEFAULT_COLLECTION = "_default"
DEFAULT_SEARCH_INDEX = "strands-memory-index"
DEFAULT_CONTENT_FIELD = "content"
DEFAULT_VECTOR_FIELD = "embedding"
DEFAULT_METADATA_FIELD = "metadata"
DEFAULT_NAMESPACE_FIELD = "namespace"
DEFAULT_NAMESPACE = "default"
DEFAULT_MAX_RESULTS = 5

JsonMap = dict[str, Any]
EmbeddingCallable = Callable[[str], list[float] | Awaitable[list[float]]]


class EmbeddingProvider(Protocol):
    """Provider that converts text into an embedding vector.

    The connector accepts any object with this method, so applications can use
    OpenAI, Bedrock, Cohere, local sentence-transformers, Capella Model Services,
    or any other embedding service without coupling this package to a vendor.
    """

    def embed(self, text: str) -> list[float] | Awaitable[list[float]]:
        """Return an embedding vector for text."""
        ...


class CouchbaseMemoryStoreConfig(MemoryStoreConfig, total=False):
    """Configuration for :class:`CouchbaseMemoryStore`."""

    connection_string: str
    username: str
    password: str
    bucket_name: str
    scope_name: str
    collection_name: str
    search_index_name: str
    content_field: str
    vector_field: str
    metadata_field: str
    namespace_field: str
    namespace: str
    dimensions: int
    embedding_provider: EmbeddingProvider | EmbeddingCallable
    cluster: Cluster
    collection: Collection
    backend: CouchbaseBackend
    num_candidates: int


class MemoryDocument(TypedDict):
    """Stored Couchbase memory document shape."""

    content: str
    embedding: list[float]
    metadata: JsonMap
    namespace: str
    created_at: str
    updated_at: str


@dataclass(frozen=True)
class SearchHit:
    """Backend-neutral search hit returned by CouchbaseBackend."""

    id: str
    score: float | None
    content: str
    metadata: JsonMap
    namespace: str | None = None


class CouchbaseBackend(Protocol):
    """Small backend protocol used by unit tests and the SDK adapter."""

    async def upsert(self, key: str, document: MemoryDocument) -> None:
        """Store or replace a memory document."""
        ...

    async def vector_search(
        self,
        *,
        search_index_name: str,
        vector_field: str,
        query_vector: list[float],
        limit: int,
        num_candidates: int | None,
        namespace: str,
        namespace_field: str,
        content_field: str,
        metadata_field: str,
    ) -> list[SearchHit]:
        """Run a vector search and return normalized hits."""
        ...

    async def close(self) -> None:
        """Close network resources if the backend owns them."""
        ...


class CouchbaseSdkBackend:
    """Couchbase Python SDK adapter.

    It uses KV for writes and Couchbase Search `VectorQuery` / `VectorSearch`
    for semantic recall. Search indexes are intentionally not created by this
    class because Couchbase vector index definitions are deployment- and
    dimension-specific. The README and setup docs show how to create them.
    """

    def __init__(
        self,
        *,
        connection_string: str,
        username: str,
        password: str,
        bucket_name: str,
        scope_name: str,
        collection_name: str,
        cluster: Cluster | None = None,
        collection: Collection | None = None,
    ) -> None:
        self._owns_cluster = cluster is None
        self._cluster = cluster or Cluster.connect(
            connection_string,
            ClusterOptions(PasswordAuthenticator(username, password)),
        )
        self._bucket = self._cluster.bucket(bucket_name)
        self._scope = self._bucket.scope(scope_name) if scope_name != DEFAULT_SCOPE else self._bucket.default_scope()
        self._collection = collection or (
            self._scope.collection(collection_name)
            if collection_name != DEFAULT_COLLECTION
            else self._bucket.default_collection()
        )

    async def upsert(self, key: str, document: MemoryDocument) -> None:
        await asyncio.to_thread(self._collection.upsert, key, document)

    async def vector_search(
        self,
        *,
        search_index_name: str,
        vector_field: str,
        query_vector: list[float],
        limit: int,
        num_candidates: int | None,
        namespace: str,
        namespace_field: str,
        content_field: str,
        metadata_field: str,
    ) -> list[SearchHit]:
        def _search() -> list[SearchHit]:
            prefilter = couchbase_search.MatchQuery(namespace, field=namespace_field)
            vector_query = VectorQuery.create(
                vector_field,
                query_vector,
                num_candidates=num_candidates or max(limit * 3, limit),
                prefilter=prefilter,
            )
            vector_search = VectorSearch.from_vector_query(vector_query)
            request = couchbase_search.SearchRequest.create(vector_search)
            result = self._scope.search(
                search_index_name,
                request,
                SearchOptions(limit=limit, fields=[content_field, metadata_field, namespace_field]),
            )
            hits: list[SearchHit] = []
            for row in result.rows():
                fields = dict(getattr(row, "fields", None) or {})
                metadata = fields.get(metadata_field) or {}
                if not isinstance(metadata, dict):
                    metadata = {"value": metadata}
                content = fields.get(content_field, "")
                hits.append(
                    SearchHit(
                        id=str(getattr(row, "id", "")),
                        score=cast(float | None, getattr(row, "score", None)),
                        content=str(content),
                        metadata=metadata,
                        namespace=cast(str | None, fields.get(namespace_field)),
                    )
                )
            return hits

        return await asyncio.to_thread(_search)

    async def close(self) -> None:
        if self._owns_cluster:
            close = getattr(self._cluster, "close", None)
            if close is not None:
                await asyncio.to_thread(close)


class CouchbaseMemoryStore(MemoryStore):
    """Couchbase Vector Search implementation of the Strands MemoryStore protocol."""

    name: str
    description: str | None
    max_search_results: int | None
    writable: bool
    extraction: Any

    def __init__(self, **store_config: Unpack[CouchbaseMemoryStoreConfig]) -> None:
        self.name = store_config.get("name", DEFAULT_NAME)
        self.description = store_config.get("description", DEFAULT_DESCRIPTION)
        self.max_search_results = store_config.get("max_search_results", DEFAULT_MAX_RESULTS)
        self.writable = store_config.get("writable", True)
        self.extraction = store_config.get("extraction")
        self.content_field = store_config.get("content_field", DEFAULT_CONTENT_FIELD)
        self.vector_field = store_config.get("vector_field", DEFAULT_VECTOR_FIELD)
        self.metadata_field = store_config.get("metadata_field", DEFAULT_METADATA_FIELD)
        self.namespace_field = store_config.get("namespace_field", DEFAULT_NAMESPACE_FIELD)
        self.namespace = store_config.get("namespace") or os.getenv("COUCHBASE_NAMESPACE") or DEFAULT_NAMESPACE
        self.search_index_name = (
            store_config.get("search_index_name") or os.getenv("COUCHBASE_SEARCH_INDEX") or DEFAULT_SEARCH_INDEX
        )
        self.dimensions = store_config.get("dimensions")
        self.num_candidates = store_config.get("num_candidates")
        provider = store_config.get("embedding_provider")
        if provider is None:
            raise ValueError("embedding_provider is required")
        self.embedding_provider = provider
        backend = store_config.get("backend")
        self._backend = backend or CouchbaseSdkBackend(
            connection_string=store_config.get("connection_string")
            or os.getenv("COUCHBASE_CONNECTION_STRING")
            or DEFAULT_CONNECTION_STRING,
            username=store_config.get("username") or os.getenv("COUCHBASE_USERNAME") or "",
            password=store_config.get("password") or os.getenv("COUCHBASE_PASSWORD") or "",
            bucket_name=store_config.get("bucket_name") or os.getenv("COUCHBASE_BUCKET") or DEFAULT_BUCKET,
            scope_name=store_config.get("scope_name") or os.getenv("COUCHBASE_SCOPE") or DEFAULT_SCOPE,
            collection_name=store_config.get("collection_name")
            or os.getenv("COUCHBASE_COLLECTION")
            or DEFAULT_COLLECTION,
            cluster=store_config.get("cluster"),
            collection=store_config.get("collection"),
        )

    async def search(self, query: str, options: StrandsSearchOptions | None = None) -> list[MemoryEntry]:
        """Search memories by vector similarity, ordered by Couchbase relevance."""
        if not query.strip():
            return []
        limit = (options or {}).get("max_search_results") or self.max_search_results or DEFAULT_MAX_RESULTS
        query_vector = await self._embed(query)
        hits = await self._backend.vector_search(
            search_index_name=self.search_index_name,
            vector_field=self.vector_field,
            query_vector=query_vector,
            limit=limit,
            num_candidates=self.num_candidates,
            namespace=self.namespace,
            namespace_field=self.namespace_field,
            content_field=self.content_field,
            metadata_field=self.metadata_field,
        )
        entries: list[MemoryEntry] = []
        for hit in hits:
            metadata = dict(hit.metadata)
            metadata.setdefault("id", hit.id)
            metadata.setdefault("score", hit.score)
            metadata.setdefault("namespace", hit.namespace or self.namespace)
            entries.append(MemoryEntry(content=hit.content, metadata=metadata))
        return entries

    async def add(self, content: str, metadata: Mapping[str, Any] | None = None) -> str:
        """Store one memory document and return its Couchbase document key."""
        if not self.writable:
            raise RuntimeError(f"Memory store {self.name!r} is not writable")
        if not content.strip():
            raise ValueError("content must not be empty")
        clean_metadata = dict(metadata or {})
        key = str(clean_metadata.pop("id", clean_metadata.pop("memory_id", f"memory::{self.namespace}::{uuid4()}")))
        vector = await self._embed(content)
        now = datetime.now(timezone.utc).isoformat()
        document: MemoryDocument = {
            "content": content,
            "embedding": vector,
            "metadata": clean_metadata,
            "namespace": self.namespace,
            "created_at": now,
            "updated_at": now,
        }
        await self._backend.upsert(key, document)
        return key

    async def add_messages(self, messages: list[Message], context: AddMessagesContext | None = None) -> list[str]:
        """Store raw conversation turns as discrete memory entries.

        Vector databases do not do server-side extraction. This method preserves
        role/content structure for callers that opt into raw message storage.
        For model-distilled facts, configure Strands extraction to use `add`.
        """
        keys: list[str] = []
        sequence_numbers = context.sequence_numbers if context else None
        for index, message in enumerate(messages):
            content = _message_to_text(message)
            if not content:
                continue
            role = str(message.get("role", "unknown"))
            metadata: JsonMap = {"role": role, "source": "strands.add_messages"}
            if sequence_numbers and index < len(sequence_numbers):
                metadata["sequence_number"] = sequence_numbers[index]
            keys.append(await self.add(content, metadata))
        return keys

    async def close(self) -> None:
        """Close backend resources."""
        await self._backend.close()

    async def _embed(self, text: str) -> list[float]:
        provider = self.embedding_provider
        if callable(provider) and not hasattr(provider, "embed"):
            result = provider(text)
        else:
            result = cast(EmbeddingProvider, provider).embed(text)
        if inspect.isawaitable(result):
            result = await result
        vector = [float(value) for value in result]
        if not vector:
            raise ValueError("embedding provider returned an empty vector")
        if self.dimensions is not None and len(vector) != self.dimensions:
            raise ValueError(f"embedding provider returned {len(vector)} dimensions; expected {self.dimensions}")
        return vector


def _message_to_text(message: Message) -> str:
    parts: list[str] = []
    for block in message.get("content", []):
        if isinstance(block, dict) and isinstance(block.get("text"), str):
            parts.append(block["text"])
    return "\n".join(parts).strip()

File: typescript/src/memory-store.ts
import * as couchbase from 'couchbase'
import type {
  AddMessagesContext,
  ExtractionConfig,
  JSONValue,
  MemoryEntry,
  MemoryStore,
  MemoryStoreConfig,
  MessageData,
  SearchOptions,
} from '@strands-agents/sdk'

const DEFAULT_NAME = 'couchbase'
const DEFAULT_DESCRIPTION = 'Long-term semantic memory stored in Couchbase Hyperscale Vector Search.'
const DEFAULT_CONNECTION_STRING = 'couchbase://localhost'
const DEFAULT_BUCKET = 'strands_memory'
const DEFAULT_SCOPE = '_default'
const DEFAULT_COLLECTION = '_default'
const DEFAULT_SEARCH_INDEX = 'strands-memory-search-index'
const DEFAULT_VECTOR_BACKEND = 'hyperscale'
const DEFAULT_DISTANCE_METRIC = 'L2_SQUARED'
const DEFAULT_CONTENT_FIELD = 'content'
const DEFAULT_VECTOR_FIELD = 'embedding'
const DEFAULT_METADATA_FIELD = 'metadata'
const DEFAULT_NAMESPACE_FIELD = 'namespace'
const DEFAULT_NAMESPACE = 'default'
const DEFAULT_MAX_RESULTS = 5

export type CouchbaseVectorBackend = 'hyperscale' | 'search'

export interface EmbeddingProvider {
  embed(text: string): Promise<number[]> | number[]
}

export interface MemoryDocument {
  content: string
  embedding: number[]
  metadata: Record<string, JSONValue>
  namespace: string
  created_at: string
  updated_at: string
}

export interface SearchHit {
  id: string
  score?: number | undefined
  content: string
  metadata: Record<string, JSONValue>
  namespace?: string | undefined
}

export interface CouchbaseBackend {
  upsert(key: string, document: MemoryDocument): Promise<void>
  vectorSearch(input: {
    searchIndexName: string
    vectorBackend: CouchbaseVectorBackend
    distanceMetric: string
    vectorField: string
    queryVector: number[]
    limit: number
    numCandidates?: number | undefined
    namespace: string
    namespaceField: string
    contentField: string
    metadataField: string
  }): Promise<SearchHit[]>
  close(): Promise<void>
}

export interface CouchbaseMemoryStoreConfig extends MemoryStoreConfig {
  connectionString?: string
  username?: string
  password?: string
  bucketName?: string
  scopeName?: string
  collectionName?: string
  searchIndexName?: string
  vectorBackend?: CouchbaseVectorBackend
  distanceMetric?: string
  contentField?: string
  vectorField?: string
  metadataField?: string
  namespaceField?: string
  namespace?: string
  dimensions?: number
  embeddingProvider: EmbeddingProvider | ((text: string) => Promise<number[]> | number[])
  backend?: CouchbaseBackend
  cluster?: couchbase.Cluster | undefined
  collection?: couchbase.Collection | undefined
  numCandidates?: number
}

export class CouchbaseSdkBackend implements CouchbaseBackend {
  private readonly config: {
    connectionString: string
    username: string
    password: string
    bucketName: string
    scopeName: string
    collectionName: string
    cluster?: couchbase.Cluster | undefined
    collection?: couchbase.Collection | undefined
  }
  private cluster: couchbase.Cluster | undefined
  private scope: couchbase.Scope | undefined
  private collection: couchbase.Collection | undefined
  private connectPromise: Promise<couchbase.Cluster> | undefined
  private readonly ownsCluster: boolean

  constructor(config: {
    connectionString: string
    username: string
    password: string
    bucketName: string
    scopeName: string
    collectionName: string
    cluster?: couchbase.Cluster | undefined
    collection?: couchbase.Collection | undefined
  }) {
    this.config = config
    this.ownsCluster = config.cluster === undefined
    if (config.cluster !== undefined) {
      this.cluster = config.cluster
      this.resolveBucketHandles(config.cluster)
    }
    if (config.collection !== undefined) {
      this.collection = config.collection
    }
  }

  async upsert(key: string, document: MemoryDocument): Promise<void> {
    const collection = await this.getCollection()
    await collection.upsert(key, document)
  }

  async vectorSearch(input: {
    searchIndexName: string
    vectorBackend: CouchbaseVectorBackend
    distanceMetric: string
    vectorField: string
    queryVector: number[]
    limit: number
    numCandidates?: number | undefined
    namespace: string
    namespaceField: string
    contentField: string
    metadataField: string
  }): Promise<SearchHit[]> {
    if (input.vectorBackend === 'search') return this.searchServiceVectorSearch(input)
    return this.hyperscaleVectorSearch(input)
  }

  async close(): Promise<void> {
    const cluster = this.cluster ?? (this.connectPromise ? await this.connectPromise : undefined)
    if (this.ownsCluster && cluster !== undefined) {
      await cluster.close()
    }
  }

  private async hyperscaleVectorSearch(input: {
    vectorField: string
    queryVector: number[]
    limit: number
    namespace: string
    namespaceField: string
    contentField: string
    metadataField: string
    distanceMetric: string
    numCandidates?: number | undefined
  }): Promise<SearchHit[]> {
    const scope = await this.getScope()
    const collection = quoteIdentifier(this.config.collectionName)
    const contentExpr = quotePath(input.contentField)
    const metadataExpr = quotePath(input.metadataField)
    const namespaceExpr = quotePath(input.namespaceField)
    const vectorExpr = quotePath(input.vectorField)
    const distanceMetricLiteral = quoteStringLiteral(validateDistanceMetric(input.distanceMetric))
    const centroidsToProbe = Math.trunc(input.numCandidates ?? 8)
    const statement = `
      SELECT META().id AS id,
             ${contentExpr} AS content,
             ${metadataExpr} AS metadata,
             ${namespaceExpr} AS namespace_value,
             APPROX_VECTOR_DISTANCE(
               ${vectorExpr},
               $query_vector,
               ${distanceMetricLiteral},
               ${centroidsToProbe}
             ) AS distance
      FROM ${collection}
      WHERE ${namespaceExpr} = $namespace
      ORDER BY distance
      LIMIT ${Math.trunc(input.limit)}
    `
    const result = await scope.query(statement, {
      parameters: {
        query_vector: input.queryVector,
        namespace: input.namespace,
      },
      scanConsistency: couchbase.QueryScanConsistency.RequestPlus,
    })
    return result.rows.map((row: any) => {
      const rawMetadata = row.metadata as JSONValue | undefined
      const metadata = isRecord(rawMetadata) ? rawMetadata : rawMetadata === undefined ? {} : { value: rawMetadata }
      return {
        id: String(row.id ?? ''),
        score: typeof row.distance === 'number' ? row.distance : undefined,
        content: String(row.content ?? ''),
        metadata,
        namespace: typeof row.namespace_value === 'string' ? row.namespace_value : undefined,
      }
    })
  }

  private async searchServiceVectorSearch(input: {
    searchIndexName: string
    vectorField: string
    queryVector: number[]
    limit: number
    numCandidates?: number | undefined
    namespace: string
    namespaceField: string
    contentField: string
    metadataField: string
  }): Promise<SearchHit[]> {
    const scope = await this.getScope()
    const prefilter = couchbase.SearchQuery.match(input.namespace).field(input.namespaceField)
    const vectorQuery = couchbase.VectorQuery.create(input.vectorField, input.queryVector)
      .numCandidates(input.numCandidates ?? Math.max(input.limit * 3, input.limit))
      .prefilter(prefilter)
    const request = couchbase.SearchRequest.create(couchbase.VectorSearch.fromVectorQuery(vectorQuery))
    const result = await scope.search(input.searchIndexName, request, {
      limit: input.limit,
      fields: [input.contentField, input.metadataField, input.namespaceField],
    })
    return result.rows.map((row: any) => {
      const fields = (row.fields ?? {}) as Record<string, JSONValue>
      const rawMetadata = fields[input.metadataField]
      const metadata = isRecord(rawMetadata) ? rawMetadata : rawMetadata === undefined ? {} : { value: rawMetadata }
      const namespace = fields[input.namespaceField]
      return {
        id: String(row.id ?? ''),
        score: typeof row.score === 'number' ? row.score : undefined,
        content: String(fields[input.contentField] ?? ''),
        metadata,
        namespace: typeof namespace === 'string' ? namespace : undefined,
      }
    })
  }

  private async getScope(): Promise<couchbase.Scope> {
    if (this.scope !== undefined) return this.scope
    const cluster = await this.getCluster()
    this.resolveBucketHandles(cluster)
    if (this.scope === undefined) throw new Error('Failed to resolve Couchbase scope')
    return this.scope
  }

  private async getCollection(): Promise<couchbase.Collection> {
    if (this.collection !== undefined) return this.collection
    await this.getScope()
    if (this.collection === undefined) throw new Error('Failed to resolve Couchbase collection')
    return this.collection
  }

  private async getCluster(): Promise<couchbase.Cluster> {
    if (this.cluster !== undefined) return this.cluster
    this.connectPromise ??= couchbase.connect(this.config.connectionString, {
      username: this.config.username,
      password: this.config.password,
    })
    this.cluster = await this.connectPromise
    return this.cluster
  }

  private resolveBucketHandles(cluster: couchbase.Cluster): void {
    const bucket = cluster.bucket(this.config.bucketName)
    this.scope = this.config.scopeName === DEFAULT_SCOPE ? bucket.defaultScope() : bucket.scope(this.config.scopeName)
    this.collection =
      this.config.collection ??
      (this.config.collectionName === DEFAULT_COLLECTION
        ? bucket.defaultCollection()
        : this.scope.collection(this.config.collectionName))
  }
}

export class CouchbaseMemoryStore implements MemoryStore {
  readonly name: string
  readonly description?: string
  readonly maxSearchResults: number
  readonly writable: boolean
  readonly extraction?: boolean | ExtractionConfig

  private readonly searchIndexName: string
  private readonly vectorBackend: CouchbaseVectorBackend
  private readonly distanceMetric: string
  private readonly contentField: string
  private readonly vectorField: string
  private readonly metadataField: string
  private readonly namespaceField: string
  private readonly namespace: string
  private readonly dimensions: number | undefined
  private readonly embeddingProvider: EmbeddingProvider | ((text: string) => Promise<number[]> | number[])
  private readonly backend: CouchbaseBackend
  private readonly numCandidates: number | undefined

  constructor(config: CouchbaseMemoryStoreConfig) {
    this.name = config.name ?? DEFAULT_NAME
    this.description = config.description ?? DEFAULT_DESCRIPTION
    this.maxSearchResults = config.maxSearchResults ?? DEFAULT_MAX_RESULTS
    this.writable = config.writable ?? true
    if (config.extraction !== undefined) this.extraction = config.extraction
    this.vectorBackend =
      config.vectorBackend ??
      (process.env.COUCHBASE_VECTOR_BACKEND as CouchbaseVectorBackend | undefined) ??
      DEFAULT_VECTOR_BACKEND
    if (this.vectorBackend !== 'hyperscale' && this.vectorBackend !== 'search') {
      throw new Error("vectorBackend must be 'hyperscale' or 'search'")
    }
    this.distanceMetric = (
      config.distanceMetric ??
      process.env.COUCHBASE_DISTANCE_METRIC ??
      DEFAULT_DISTANCE_METRIC
    ).toUpperCase()
    this.searchIndexName = config.searchIndexName ?? process.env.COUCHBASE_SEARCH_INDEX ?? DEFAULT_SEARCH_INDEX
    this.contentField = config.contentField ?? DEFAULT_CONTENT_FIELD
    this.vectorField = config.vectorField ?? DEFAULT_VECTOR_FIELD
    this.metadataField = config.metadataField ?? DEFAULT_METADATA_FIELD
    this.namespaceField = config.namespaceField ?? DEFAULT_NAMESPACE_FIELD
    this.namespace = config.namespace ?? process.env.COUCHBASE_NAMESPACE ?? DEFAULT_NAMESPACE
    this.dimensions = config.dimensions
    this.embeddingProvider = config.embeddingProvider
    this.numCandidates = config.numCandidates
    this.backend =
      config.backend ??
      new CouchbaseSdkBackend({
        connectionString:
          config.connectionString ?? process.env.COUCHBASE_CONNECTION_STRING ?? DEFAULT_CONNECTION_STRING,
        username: config.username ?? process.env.COUCHBASE_USERNAME ?? '',
        password: config.password ?? process.env.COUCHBASE_PASSWORD ?? '',
        bucketName: config.bucketName ?? process.env.COUCHBASE_BUCKET ?? DEFAULT_BUCKET,
        scopeName: config.scopeName ?? process.env.COUCHBASE_SCOPE ?? DEFAULT_SCOPE,
        collectionName: config.collectionName ?? process.env.COUCHBASE_COLLECTION ?? DEFAULT_COLLECTION,
        cluster: config.cluster,
        collection: config.collection,
      })
  }

  async search(query: string, options?: SearchOptions): Promise<MemoryEntry[]> {
    if (query.trim().length === 0) return []
    const limit = options?.maxSearchResults ?? this.maxSearchResults ?? DEFAULT_MAX_RESULTS
    const queryVector = await this.embed(query)
    const hits = await this.backend.vectorSearch({
      searchIndexName: this.searchIndexName,
      vectorBackend: this.vectorBackend,
      distanceMetric: this.distanceMetric,
      vectorField: this.vectorField,
      queryVector,
      limit,
      numCandidates: this.numCandidates,
      namespace: this.namespace,
      namespaceField: this.namespaceField,
      contentField: this.contentField,
      metadataField: this.metadataField,
    })
    return hits.map((hit) => ({
      content: hit.content,
      metadata: {
        ...hit.metadata,
        id: hit.metadata.id ?? hit.id,
        score: hit.metadata.score ?? hit.score ?? null,
        namespace: hit.metadata.namespace ?? hit.namespace ?? this.namespace,
      },
    }))
  }

  async add(content: string, metadata?: Record<string, JSONValue>): Promise<string> {
    if (!this.writable) throw new Error(`Memory store ${this.name} is not writable`)
    if (content.trim().length === 0) throw new Error('content must not be empty')
    const cleanMetadata: Record<string, JSONValue> = { ...(metadata ?? {}) }
    const metadataId = cleanMetadata.id ?? cleanMetadata.memory_id
    delete cleanMetadata.id
    delete cleanMetadata.memory_id
    const key = typeof metadataId === 'string' ? metadataId : `memory::${this.namespace}::${crypto.randomUUID()}`
    const embedding = await this.embed(content)
    const now = new Date().toISOString()
    await this.backend.upsert(key, {
      content,
      embedding,
      metadata: cleanMetadata,
      namespace: this.namespace,
      created_at: now,
      updated_at: now,
    })
    return key
  }

  async addMessages(messages: MessageData[], context?: AddMessagesContext): Promise<string[]> {
    const keys: string[] = []
    for (const [index, message] of messages.entries()) {
      const content = messageToText(message)
      if (!content) continue
      const metadata: Record<string, JSONValue> = {
        role: message.role ?? 'unknown',
        source: 'strands.addMessages',
      }
      const sequenceNumber = context?.sequenceNumbers?.[index]
      if (sequenceNumber !== undefined) metadata.sequence_number = sequenceNumber
      keys.push(await this.add(content, metadata))
    }
    return keys
  }

  async close(): Promise<void> {
    await this.backend.close()
  }

  private async embed(text: string): Promise<number[]> {
    const result =
      typeof this.embeddingProvider === 'function' ? this.embeddingProvider(text) : this.embeddingProvider.embed(text)
    const vector = (await result).map((value) => Number(value))
    if (vector.length === 0) throw new Error('embedding provider returned an empty vector')
    if (this.dimensions !== undefined && vector.length !== this.dimensions) {
      throw new Error(`embedding provider returned ${vector.length} dimensions; expected ${this.dimensions}`)
    }
    return vector
  }
}

function messageToText(message: MessageData): string {
  return (message.content ?? [])
    .map((block: any) => (typeof block?.text === 'string' ? block.text : ''))
    .filter(Boolean)
    .join('\n')
    .trim()
}

function validateDistanceMetric(distanceMetric: string): string {
  const metric = distanceMetric.toUpperCase()
  const allowed = new Set(['COSINE', 'DOT', 'L2', 'EUCLIDEAN', 'L2_SQUARED', 'EUCLIDEAN_SQUARED'])
  if (!allowed.has(metric)) throw new Error(`distanceMetric must be one of ${[...allowed].join(', ')}`)
  return metric
}

function quoteStringLiteral(value: string): string {
  return `'${value.replaceAll("'", "''")}'`
}

function quoteIdentifier(identifier: string): string {
  if (!identifier || identifier.includes('\0')) throw new Error('SQL++ identifiers must be non-empty strings')
  return `\`${identifier.replaceAll('`', '``')}\``
}

function quotePath(path: string): string {
  return path.split('.').map(quoteIdentifier).join('.')
}

function isRecord(value: unknown): value is Record<string, JSONValue> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}
